### ISSUE TITLE
gunicorn fastapi  use loguru

### ISSUE THREAD

**Issue by fanmaoyu0871** [Delgan/loguru#1111]
I used this command to deploy the fastapi `gunicorn main:app -b 0.0.0.0:12000 -w 4 -k uvicorn.workers.UvicornWorker --daemon` ，Logoru was used to record HTTP requests and responses, but I found that some logs were not recorded，my loguru config：`logger.add("logs/api.log", encoding="utf-8", catch=True, backtrace=True, diagnose=True, enqueue=True, rotation="00:00", retention="20 days", compression="zip")`，is `-w 4` or `-k uvicorn.workers.UvicornWorker` has some problems? 
use `-w 1`，it's likely work fine。

**Comment by Delgan** [comment]
Yes, this is because of the multiple processes started by FastAPI. They won't inherit the `logger`, therefore the `enqueue=True` parameter is unfortunately ineffective.

Right now, it's advised to use one unique file per worker, until a new workaround is implemented in Loguru.

**Comment by fanmaoyu0871** [comment]
How to do it? Can you give me some ideas? Thank you

**Comment by 4linuxfun** [comment]
I hope this works for you.I'm using it. https://gist.github.com/nkhitrov/a3e31cfcc1b19cba8e1b626276148c49

**Comment by Delgan** [comment]
Maybe these snippets can help you:
* https://github.com/tiangolo/fastapi/issues/1276#issuecomment-615877177
* https://medium.com/1mgofficial/how-to-override-uvicorn-logger-in-fastapi-using-loguru-124133cdcd4e

**Comment by Lizidong01** [comment]
Are there any new developments on this issue?

**Comment by Delgan** [comment]
After revisiting this ticket, I finally think combining Loguru and Gunicorn should work fine, provided the configuration is set up correctly.

It turns out that Gunicorn starts its workers using the fork method, so child processes inherit the memory of the parent process. Therefore, it is critical that the logger is configured exactly once in the master process, especially when `enqueue=True`.

This is possible by using a server hook, specifically [`on_starting`](https://docs.gunicorn.org/en/latest/settings.html#on-starting). Here is a demonstration:

```python
# app.py
from fastapi import FastAPI
from loguru import logger

app = FastAPI()


@app.get("/")
def read_root():
    logger.info("Root endpoint accessed")
    return {"message": "Hello, World!"}
```
```python
# conf.py
from loguru import logger
import sys


def on_starting(server):
    logger.remove()
    logger.add(sys.stderr, enqueue=True)
    logger.add(
        "logs/api.log",
        encoding="utf-8",
        catch=True,
        backtrace=True,
        diagnose=True,
        enqueue=True,
        rotation="00:00",
        retention="20 days",
        compression="zip",
    )
```
```bash
# Usage
gunicorn app:app -b 0.0.0.0:12000 -w 4 -k uvicorn.workers.UvicornWorker -c conf.py
```

When using this solution, workers will automatically inherit the logger installed on the server, which will ensure that logs are correctly synchronized across different processes. It is essential that the logger is not configured once per worker, as this would imply that the handlers are duplicated, causing problems when rotating the log file, for example.

If you are still experiencing difficulties, please feel free to create a new ticket with a minimal example so that I can reproduce your configuration and try to resolve your issue.
